[RNMobile] Embed Block Bottom Sheet to set a link

ceyhun opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

This PR implements the Embed Block Bottom Sheet to set a link.

Gutenberg Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3436

### Use of `onClose` and `onDismiss` callbacks from the bottom sheet component

It's not very clear the difference between the `onClose` and `onDismiss` callbacks, so for the purpose of this PR, I'd like to briefly describe when these callbacks are triggered and how they're used in the embed sheet.

#### `onClose`
This callback is triggered by the actions listed below, as far as I checked, looks like the callback is intended for requesting the parent to handle the close flow.

One thing I'd like to note is that this being called multiple times in some cases, I'm not sure if this behavior is intended or not, so it would be nice that we investigate it deeper in the future. As an example, when tapping the backdrop element when the modal is visible, the callback is triggered a first time, and then right after, as the modal is dismissed, it's triggered for a second time.

This is caused by the `onDismiss` function from the component that for some reason, calls `onCloseBottomSheet` that triggers the `onClose` callback:
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L310-L318
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L340-L351

**Actions that trigger `onClose` callback:**
- Tapping the backdrop element (related to [`onBackdropPress` callback](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L497)).
- Pressing the back button (only Android) (related to [`onBackButtonPress` callback](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L498)).
- Swiping down the modal (related to [`onSwipeComplete callback`](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L499)).
- Perform the escape gesture (related to [`onAccessibilityEscape` callback](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L513)).
- After the modal is dismissed ([related code](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/components/src/mobile/bottom-sheet/index.native.js#L310-L318)).

#### `onDismiss`
This is triggered when the modal has been already dismissed/hidden, which only happens once in the closing flow.

#### How are they used in the embed sheet

- `onClose`: As is intended for hiding the modal, we're using it just to update the internal state of the placeholder component and set the visibility value to `false`. As mentioned earlier, this can be triggered multiple times but since we're only updating the internal state this is not an issue.
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-library/src/embed/embed-placeholder.native.js#L80
- `onDismiss`: It's being used for handling the URL attribute update. The reason for using this callback instead of `onClose` is because we have the guarantee that will be called only once.
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-library/src/embed/embed-bottom-sheet.native.js#L36-L46

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Add the Embed block
2. Bottom sheet will open and link field will be focused
3. If there is a link in the clipboard it should be auto-pasted
4. Type a link if empty
5. Can click on the “Learn more about embeds” link at the bottom to open the support article in the browser
6. Close the bottom sheet
7. If the link is not a valid URL, a warning message is shown

## Missing features

- ~~Can re-open by pressing on the embed block placeholder. 
Depends on https://github.com/wordpress-mobile/gutenberg-mobile/issues/3273.~~
- Can open by pressing the pen icon in the toolbar when an existing embed block is selected. 
Depends on https://github.com/wordpress-mobile/gutenberg-mobile/issues/3276 
- Edit/replace the current URL placeholder, this will be addressed on https://github.com/wordpress-mobile/gutenberg-mobile/issues/3285

## Screenshots <!-- if applicable -->

iOS Light Mode | iOS Dark Mode 
--------|-------
<img width="300" src="https://user-images.githubusercontent.com/14905380/123625333-f3b88f80-d80f-11eb-988d-ab644ba35f04.png">|<img width="300" src="https://user-images.githubusercontent.com/14905380/123625338-f5825300-d80f-11eb-93fa-0d8fd4ffa27a.png">

Android Light Mode | Android Dark Mode 
--------|-------
<img width="300" src="https://user-images.githubusercontent.com/14905380/123625778-688bc980-d810-11eb-8196-6dac86449137.jpg">|<img width="300" src="https://user-images.githubusercontent.com/14905380/123625787-6aee2380-d810-11eb-9973-442e8a44f77c.jpg">

iOS video capture|Android video capture
--|--
<video src=https://user-images.githubusercontent.com/14905380/123624724-362d9c80-d80f-11eb-9005-6182a843235c.mp4>|<video src=https://user-images.githubusercontent.com/14905380/123626445-2adb7080-d811-11eb-93a8-a8a35835482c.mp4>

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ceyhun committed:
Add separatorType option to LinkSettings footer

ceyhun committed:
Create native EmbedBottomSheet component

ceyhun committed:
Use EmbedBottomSheet in native EmbedPlaceholder

jd-alexander commented on packages/block-library/src/embed/embed-bottom-sheet.native.js:
```diff
@@ -0,0 +1,60 @@
+	const [ url, setUrl ] = useState( value );
+	const { createErrorNotice } = useDispatch( noticesStore );
+
+	function onCloseWithUrl() {
```
@ceyhun I am noticing that when the bottom sheet is opened and then closed the error below is shown while the Embed link is empty. Is that the expected behavior or do we want to check if the URL is empty or not before checking `isURL`?

fluiddot replied:
I decided to change the callback we use for returning the URL to `setAttributes` because `onClose` was being called multiple times and each time has a different URL value:

Example URL: `wordpress.com`
- 1st call: `url = wordpress.com`
- 2nd call: `url = http://wordpress.com`

**Related commit: https://github.com/WordPress/gutenberg/pull/31343/commits/2db674c7c59aeb3b54e14a1930a7a18689635682**

jd-alexander commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -8,12 +8,25 @@ import { TextInput } from 'react-native';
+				value={ value }
+				placeholder={ __( 'Enter URL to embed here…' ) }
+			/>
+			<EmbedBottomSheet
```
Since the https://github.com/WordPress/gutenberg/pull/30592 has been merged in, I will give this PR another review when the `EmbedBottomSheet` is migrated there.

fluiddot replied:
I've just updated the PR with `trunk` (which includes the changes from #30592) and applied some updates.

fluiddot committed:
Merge branch 'trunk' into rnmobile/embed-block-bottom-sheet

# Conflicts:
#	packages/block-library/src/embed/embed-placeholder.native.js

fluiddot committed:
Add onSetAttributes callback to embed bottom sheet

When the attributes are modified in the LinkSettingsNavigation component the setAttributes callback is triggered. This callback is called once unlike the onClose callback, which is called multiple times and doesn't guarantee that the attributes are up-to-date at the time is triggered.

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -26,21 +37,37 @@ const EmbedPlaceholder = ( { icon, label, onFocus } ) => {
-			</View>
-		</TouchableWithoutFeedback>
+		<>
+			{ value ? (
+				<Text>{ value }</Text>
+			) : (
```
The block will show the current URL until we implement the preview part.

fluiddot committed:
Correct isEmbedSheetVisible typo

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -26,21 +37,37 @@ const EmbedPlaceholder = ( { icon, label, onFocus } ) => {
+				isVisible={ isEmbedSheetVisible }
+				onClose={ () => setIsEmbedSheetVisible( false ) }
+				onSetAttributes={ ( { url } ) => {
+					onChange( { target: { value: url } } );
```
The object returned here has to match what is expected in the `edit` component:
https://github.com/WordPress/gutenberg/blob/8d3db371b033d4e331702a7a837978fac79e82ca/packages/block-library/src/embed/edit.js#L214

hypest replied:
Hmm, sounds like useful context to have as inline comment instead of just a PR comment, wdyt?

hypest replied:
Addressed with https://github.com/WordPress/gutenberg/pull/31343/commits/2a337a667a49b8919cf4f1b3b954daf37c2f9d7b.

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -26,21 +37,37 @@ const EmbedPlaceholder = ( { icon, label, onFocus } ) => {
+				onClose={ () => setIsEmbedSheetVisible( false ) }
+				onSetAttributes={ ( { url } ) => {
+					onChange( { target: { value: url } } );
+					onSubmit();
```
`onSubmit` is the callback that actually updates the attributes of the block:
https://github.com/WordPress/gutenberg/blob/8d3db371b033d4e331702a7a837978fac79e82ca/packages/block-library/src/embed/edit.js#L204-L211

fluiddot replied:
The behavior of setting a URL differs a bit from the web version, there `onChange` is used to update the internal state when the user types the URL, in the native version this is done within the `EmbedBottomSheet` component so for update attributes we have to explicitly call both callbacks:
- `onChange` with the URL to update the internal state
- `onSubmit` to set the attributes with the previous updated internal state

hypest replied:
Speaking of the `onSubmit()`, looks like it's called twice when dismissing the bottom sheet. Not sure if that's what you also witness @fluiddot?

fluiddot replied:
Oh that's interesting, I didn't spot it when testing, I'll review this part deeper, thanks for the heads-up 🙇 !

fluiddot replied:
True, the `onSubmit` callback gets called twice, looks like the `onCloseBottomSheet` callback from the bottom sheet component is called twice from different places and end up setting the attributes two times:
1. `onBackdropPress` prop from the Modal:  https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/bottom-sheet/index.native.js#L496
2. `onDismiss` callback from Modal: https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/bottom-sheet/index.native.js#L316

When `onCloseBottomSheet` is called, internally it sets the attributes by the following flow:
- The `LinkSettings` components sets the `handleClosingBottomSheet` callback:  https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/link-settings/index.native.js#L101
- `onCloseBottomSheet` calls the callback `handleClosingBottomSheet` (this callback is the one previously set in the above line):
https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/bottom-sheet/index.native.js#L343
- The callback internally calls the function to set the attributes:
https://github.com/WordPress/gutenberg/blob/779fdaa94263611a3ccb5f7fdd5f312b43e5cbea/packages/components/src/mobile/link-settings/index.native.js#L177

At this point, I'm leaning towards using the `onDismiss` callback to trigger the submit as it looks the only one that would be called only once and use `onSetAttributes` to update the internal state as the web version does 🤔 .

hypest replied:
Ah, thanks for confirming Carlos! the "duplicate call" of `onSubmit` is worrying, but not sure yet if we should block this PR on it. I'll check develop.

fluiddot replied:
After [this commit](https://github.com/WordPress/gutenberg/pull/31343/commits/0154c6f0131b633e97ca79dc9a25791a3fffd6ac), now the `onSubmit` callback is not called twice because it's located in a different callback, before it was triggered in `onSetAttributes` and now is [`onDismiss`](https://github.com/WordPress/gutenberg/pull/31343/files#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fR39), so this is no longer an issue.

`onSetAttributes ` callback will still be called twice, as it's described in the previous comment, but I think it shouldn't produce any bad side effect as it's only updating the internal state of the embed sheet component ([code reference](https://github.com/WordPress/gutenberg/pull/31343/files#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fR48-R50)).

hypest replied:
> onSetAttributes callback will still be called twice, as it's described in the previous comment, but I think it shouldn't produce any bad side effect as it's only updating the internal state of the embed sheet component ([code reference](https://github.com/WordPress/gutenberg/pull/31343/files#diff-7d155d7b9eb26681fd43d964649784c4799d33b187271b3288e604f9c9728b8fR48-R50)).

I think it will also call the [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407) and update the attributes in the store (makes sense), maybe causing some extra renders. I'm not fully sure. It still doesn't sound like bad side effects, except perhaps for performance. Anyway, I think this is not a blocker for this PR. 👍

fluiddot replied:
> I think it will also call the block's onChange and update the attributes in the store (makes sense), maybe causing some extra renders. I'm not fully sure. It still doesn't sound like bad side effects, except perhaps for performance.

I'd like to note that the code displayed above is outdated, the current code only does the internal update when the `setAttributes` is triggered, so it shouldn't call [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407):
https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-library/src/embed/embed-bottom-sheet.native.js#L48-L62

I also checked the code of the component that handles the link settings ([`LinkSettings`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/link-settings/index.native.js)) just in case it's updating the attributes internally but it doesn't, it only triggers the callback. Here is the code in charge of notifying the update of the attributes:
https://github.com/WordPress/gutenberg/blob/326ebc6044418d91df2d58c0c9dce7548e031ab1/packages/components/src/mobile/link-settings/index.native.js#L161-L174

hypest replied:
>  so it shouldn't call [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407):

Hmm, I do see the line getting hit in the debugger on 32c4466f62 🤔

fluiddot replied:
> Hmm, I do see the line getting hit in the debugger on 32c4466 🤔

Oh I think I wasn't clear on my explanation, the `setAttributes` callback will be called twice but it won't trigger the [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407), instead, it will be triggered by the `onSubmit` callback when the modal is dismissed not when the [`LinkSettings`](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/mobile/link-settings/index.native.js) component notifies the update of the attributes.

Here I added a patch diff that logs in the console calls to both callbacks `setAttributes` and [block's onChange](https://github.com/WordPress/gutenberg/blob/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4/packages/block-editor/src/components/block-list/block.native.js#L407) in case it makes my explanation clearer 😄 :
```diff
diff --git a/packages/block-library/src/embed/embed-bottom-sheet.native.js b/packages/block-library/src/embed/embed-bottom-sheet.native.js
index e7e69fb994302582a9fa8fb22f10bfd64f5b9156..4e910047ea6fa97deedf591057f3c4d9f875b470 100644
--- a/packages/block-library/src/embed/embed-bottom-sheet.native.js
+++ b/packages/block-library/src/embed/embed-bottom-sheet.native.js
@@ -46,6 +46,7 @@ const EmbedBottomSheet = ( { value, isVisible, onClose, onSubmit } ) => {
 	}, [ url, onSubmit ] );
 
 	function setAttributes( attributes ) {
+		console.log( 'setAttributes', attributes );
 		setURL( attributes.url );
 	}
 
diff --git a/packages/block-editor/src/components/block-list/block.native.js b/packages/block-editor/src/components/block-list/block.native.js
index 018cf4760a07ebc6f208452b9ed01c6de0d0e57a..418cb025b260d7311284689967f3661a004e2585 100644
--- a/packages/block-editor/src/components/block-list/block.native.js
+++ b/packages/block-editor/src/components/block-list/block.native.js
@@ -404,6 +404,7 @@ export default compose( [
 				selectBlock( clientId, initialPosition );
 			},
 			onChange: ( attributes ) => {
+				console.log( 'onChange', attributes );
 				updateBlockAttributes( ownProps.clientId, attributes );
 			},
 			onReplace( blocks, indexToSelect ) {
```

As you can see `setAttributes` is called two times but `onChange` only once, so the whole block shouldn't get re-rendered, however, it will the `EmbedBottomSheet` component as its internal state is updated:

<img width="573" alt="Screenshot 2021-07-01 at 10 29 41" src="https://user-images.githubusercontent.com/14905380/124092537-420b9f80-da57-11eb-99aa-592c2baef4da.png">

fluiddot commented:
👋 @hypest The PR is ready for review, I've updated it with the latest changes from `trunk` and applied some fixes related to passing the URL value to the block , let me know if I can help with the review, thanks 🙇 !

hypest commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -8,13 +8,24 @@ import { View, Text, TouchableWithoutFeedback } from 'react-native';
+	onSubmit,
+	onChange,
+} ) => {
+	const [ isEmbedSheetVisible, setIsEmbedSheetVisible ] = useState( ! value );
```
Noticed a flow that triggers the bottom sheet to show: If you add the block in html mode (i.e. add a `<!-- wp:embed /-->` markup) and switch to visual mode, the bottom sheet gets opened automatically. Can we make the flow specific, to only auto-open when added in visual mode?

fluiddot replied:
Good catch!

> Can we make the flow specific, to only auto-open when added in visual mode?

Yeah, we could have a similar flow as we have for the media blocks that only in case the component is inserter from the inserter menu we displayed the bottom sheet.

hypest replied:
Addressed with https://github.com/WordPress/gutenberg/pull/31343/commits/0d704574ed2bd6e811628602d58b315af7f30391.

hypest commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -62,9 +60,9 @@ const EmbedPlaceholder = ( {
-					setIsBottomSheetVisible( false );
+				isVisible={ isEmbedSheeVisible }
+				onClose={ () => setIsEmbedSheetVisible( false ) }
+				onSetAttributes={ ( { url } ) => {
```
With the change from using `onClose` to using `onSetAttributes` for setting the attributes (and `onSubmit` is now in the latter), I wonder if you have any thought on which event exactly should cause the preview to reload? I mean, I know the preview is not part of this PR, but I wonder what's the assumption about the API here... would we use the onClose or the URL attribute change as the preview reload event?

fluiddot replied:
I haven't checked the preview part yet but I think it should work fine by using the `onSetAttributes` event because, as far as I checked, it's only called when the bottom sheet is closed after changing the URL. However, after reading [your comment about the `onSubmit` callback being called twice](https://github.com/WordPress/gutenberg/pull/31343#discussion_r659792077), I'll review the logic just in case I'm overlooking something. What I'm mostly sure that we shouldn't do is rely on the `onClose` event because after debugging it, I realized that is called multiple times.

fluiddot replied:
I posted a comment related to this topic [here](https://github.com/WordPress/gutenberg/pull/31343/files#r659948242), I think the event that we could use for this purpose would be `onDismiss`.

hypest replied:
Hehe, starting to feel unsure what exactly each of those "similar" props/events actually do and where they differ 🤔. I mean, `onClose` and `onDismiss` sound too close, and, which one would/should call `setAttributes` even? Is it any clearer to you at this Carlos? if so, can you add a line comment about it, and ideally include a relevant explanation in the PR description? Thanks!

fluiddot replied:
Yeah agreed, both events sound too close, and to be honest I have the same uncertainty feeling. From what I understand after checking the bottom sheet component, the main differences are:
- `onClose` is triggered when one of the following actions happen (see below), my impression is that this callback is intended for requesting the parent to handle the close flow:
    - Tapping the backdrop element (related to `onBackdropPress` callback).
    - Pressing the back button (only Android) (related to `onBackButtonPress` callback).
    - Perform the escape gesture (related to `onAccessibilityEscape` callback).
    - After the modal is dismissed ([related code](https://github.com/WordPress/gutenberg/blob/83f0f6b0cefe8e1c87a4a2edc118b61f5d5e0620/packages/components/src/mobile/bottom-sheet/index.native.js#L316))
- `onDismiss` on the other hand is triggered when the modal has been already dismissed/hidden, which only happens once in the closing flow.

Probably we should review these flows because I find it weird that `onClose` can be called twice, for example, when tapping the backdrop element (first call) and then when the modal is dismissed (second call).

fluiddot replied:
> can you add a line comment about it, and ideally include a relevant explanation in the PR description?

Ideally, we should have a `README` file for documenting the bottom sheet component, I think that would be the best place to explain how the `onClose` and `onDismiss` props work, wdyt? If you agree, I could draft a separate PR for that change.

hypest replied:
A README would indeed be a better long term place. I guess what we can surface in the meantime, in the context of this PR is the why we went with `onDismiss` for example. That the kind of code design choices that I usually want to have in the PR description to give the needed context for the reader to understand the changes the PR introduces.

fluiddot replied:
Right, makes sense 👍 .

I've just updated the PR description adding the insights of the use of both callbacks for this PR in the "Use of onClose and onDismiss callbacks from the bottom sheet component" section. Let me know if the description clarifies this topic and of course any feedback you have, thanks so much for the thorough review 🙇 .

hypest reviewed (commented):
Thanks for the updates @fluiddot and good job with the initial implementation @ceyhun!

I made a first review pass and left some comments. Generally looks good to me but I'm not sure yet about state/data updates. I probably need another, deeper pass to check if there's an issue with the bottom sheets in general or it's something specific to the implementation on this PR. Let me know if your thoughts in the comments.

fluiddot commented:
> Thanks for the updates @fluiddot and good job with the initial implementation @ceyhun!
> 
> I made a first review pass and left some comments. Generally looks good to me but I'm not sure yet about state/data updates. I probably need another, deeper pass to check if there's an issue with the bottom sheets in general or it's something specific to the implementation on this PR. Let me know if your thoughts in the comments.

Thanks @hypest for the review! I'll review further the state/data updates logic and provide further information about how it works and if the approach we're following here is reliable for the purpose of this block.

fluiddot committed:
Use onDismiss callback to notify embed URL value update

fluiddot committed:
Add comment to onChangeURL embed placeholder

fluiddot committed:
Only display the embed sheet on block insertion

fluiddot committed:
Add native version of onSubmit in embed

fluiddot committed:
Prevent onDismiss prop overwriting in bottom sheet

I noticed that if a onDismiss prop is passed to the bottom sheet component, the internal function onDismiss is not executed. This behavior looks wrong because this function is intended to actually call the onDismiss prop.

fluiddot committed:
Show only the embed sheet automatically if empty

fluiddot commented on packages/block-library/src/embed/edit.js:
```diff
@@ -192,6 +192,26 @@ const EmbedEdit = ( props ) => {
+		setAttributes( { url } );
+	};
+
+	const onSubmitNative = ( value ) => {
+		// On native, the URL change is only notified when submitting
+		// so we have to explicitly set the URL.
+		setURL( value );
+
+		// Replicate the same behavior as onSubmit
+		setIsEditingURL( false );
+		setAttributes( { url: value } );
+	};
```
I tried to replicate the web flow for dispatching URL updates and submitting but I couldn't find a reliable way to achieve it. The approach I tried to follow was using the `onChange` callback when the `setAttributes` callback of `LinkSettingsNavigation ` component is triggered. The problem is that `setAttributes` can be called two times if you tap on the submit/intro button on the keyboard, in fact, the second call is happening after dismissing the bottom sheet so it updates the URL after calling `onSubmit` which is not expected.

Here you can watch a debugging session of the described issue:

https://user-images.githubusercontent.com/14905380/123791721-10bc9380-d8e0-11eb-83d5-b827d4d0d6c9.mp4

For this reason, I had to create a native version of the `onSubmit` callback, which instead of an event it receives the URL value and does both updating the URL and submitting.

hypest replied:
Want to update the inline comment (`On native, the URL change is only notified when submitting`) to include something like `, and not via 'onChange'`? I had to debug the code to verify that the [`url` state](https://github.com/WordPress/gutenberg/blob/b5ee74ef9c968b194221e0d4e0d97dc04b479fd3/packages/block-library/src/embed/edit.js#L65) is indeed not getting updated from [this onChange](https://github.com/WordPress/gutenberg/blob/b5ee74ef9c968b194221e0d4e0d97dc04b479fd3/packages/block-library/src/embed/edit.js#L230), which is a bit unintuitive that it's not getting updated in onChange. 😬

fluiddot replied:
Good idea, I'll update it.

I always found a bit confusing the case of having a list of props but not knowing which one are only for web version and others for native. In this case, the [onChange](https://github.com/WordPress/gutenberg/blob/b5ee74ef9c968b194221e0d4e0d97dc04b479fd3/packages/block-library/src/embed/edit.js#L230) prop is only used in the web version.

hypest replied:
let's do it in a new PR? The CI tests just got all-green after multiple re-runs 😬

fluiddot replied:
Ok, let's do it that way, looks like we're lucky that all CI tests are 🟢 😅 .

fluiddot commented on packages/block-library/src/embed/edit.js:
```diff
@@ -201,21 +221,18 @@ const EmbedEdit = ( props ) => {
 					tryAgain={ () => {
 						invalidateResolution( 'getEmbedPreview', [ url ] );
 					} }
+					isSelected={ isSelected }
```
This prop was added because it's required in the calculation for showing automatically the embed sheet ([code reference](https://github.com/WordPress/gutenberg/pull/31343/files#diff-794d5b4e5171d2fc3a36373cdbb2abd0fa4dd7a5892a759891ddaf1af0563accR33-R43)).

fluiddot commented on packages/components/src/mobile/bottom-sheet/index.native.js:
```diff
@@ -393,6 +393,8 @@ class BottomSheet extends Component {
 			children,
 			withHeaderSeparator = false,
 			hasNavigation,
+			// eslint-disable-next-line no-unused-vars
+			onDismiss,
```
I noticed that if the `onDismiss` prop is passed to the bottom sheet component, the internal function `onDismiss` ([code reference](https://github.com/WordPress/gutenberg/blob/fa3bfd5e960fb6d9901edc05530a27454ecb912a/packages/components/src/mobile/bottom-sheet/index.native.js#L309-L317)) is not executed. This behavior looks wrong because this function is intended to actually call the `onDismiss` prop. To address it I added the prop to this props destructuring list to prevent including it later on when the component is rendered ([code reference](https://github.com/WordPress/gutenberg/blob/fa3bfd5e960fb6d9901edc05530a27454ecb912a/packages/components/src/mobile/bottom-sheet/index.native.js#L513)).

hypest replied:
Aha, I see what's going on there, leaving the `onDismiss` in the `rest` was ending up overriding the [attempt to set it here](https://github.com/WordPress/gutenberg/blob/fa3bfd5e960fb6d9901edc05530a27454ecb912a/packages/components/src/mobile/bottom-sheet/index.native.js#L499).

Cool trick to fix it here, though I'm worried this is very easy to break because it's a bit unintuitive and too far away from the codesite it affects. I'd recommend adding a line comment in the "extraction" point here to inform the reader what's going on.

Alternatively, what do you think about a different way to remove the `onDismiss` preset in rest closer to the code we want the removal? Like, instead of [`{ ...rest }`](https://github.com/WordPress/gutenberg/blob/fa3bfd5e960fb6d9901edc05530a27454ecb912a/packages/components/src/mobile/bottom-sheet/index.native.js#L513) wdyt about removing the onDismiss:

```javascript
{ ...( { onDismiss: _, ...rest } ) => rest } // removing the `onDismiss` as we're actually setting it earlier above and this one would override it.
```

fluiddot replied:
> Cool trick to fix it here, though I'm worried this is very easy to break because it's a bit unintuitive and too far away from the codesite it affects. I'd recommend adding a line comment in the "extraction" point here to inform the reader what's going on.

Sure! I'll add a comment to make it clearer 👍 .

> Alternatively, what do you think about a different way to remove the onDismiss preset in rest closer to the code we want the removal? Like, instead of { ...rest } wdyt about removing the onDismiss

Yeah, that would do the trick, I'm wondering if we could also use [`omit`](https://lodash.com/docs/4.17.15#omit) as we use on other places:
```
{ ...omit( rest, 'onDismiss' ) }
```

hypest replied:
> Yeah, that would do the trick, I'm wondering if we could also use omit as we use on other places:
>
> { ...omit( rest, 'onDismiss' ) }

Ah, right, we might as well use lodash like that, it's also more easy to read 👍

fluiddot replied:
I addressed this issue in [this commit](https://github.com/WordPress/gutenberg/pull/31343/commits/32c4466f62f4cb6f71e93e3256b76a1649d3b1d4) 🎊 .

fluiddot commented on packages/block-library/src/embed/embed-placeholder.native.js:
```diff
@@ -8,13 +8,40 @@ import { View, Text, TouchableWithoutFeedback } from 'react-native';
+		[ clientId ]
+	);
+	const [ isEmbedSheetVisible, setIsEmbedSheetVisible ] = useState(
+		isSelected && wasBlockJustInserted && ! value
```
This is the condition to determine if the embed sheet should be automatically displayed when the component is mounted.

hypest replied:
FWIW, verified that the bottomsheet is not triggered if we duplicate the block (via the block's toolbar 3-dot menu).

fluiddot replied:
This issue was addressed in [this commit](https://github.com/WordPress/gutenberg/pull/31343/commits/b5ee74ef9c968b194221e0d4e0d97dc04b479fd3) 🎊 .

fluiddot commented on packages/block-library/src/embed/edit.js:
```diff
@@ -201,21 +221,18 @@ const EmbedEdit = ( props ) => {
-						setIsEditingURL( false );
-						setAttributes( { url } );
-					} }
+					onSubmit={ Platform.select( {
+						web: onSubmit,
+						native: onSubmitNative,
+					} ) }
```
Although this change shouldn't break the logic of the web version, I think it's recommended to test it as a double-check.

hypest replied:
I verified the web GB using the web `onSubmit()` and only calling it once for that matter ✅

fluiddot committed:
Fix clipboard case for embed sheet

fluiddot commented:
@hypest I investigated deeper the state/data updates and I spotted some cases that weren't working properly, so I addressed them (they're posted as comments in the file changes) and refactored a bit the PR for this purpose. Being this said, the PR is ready for a new review, I'd appreciate it if you could take a look, thanks 🙇 !

hypest commented:
I'm still in the process of reviewing the new changes @fluiddot , but wanted to surface something I noticed in the meantime: Looks like the BlockSheet [onDismiss()](https://github.com/WordPress/gutenberg/blob/6da0390f864c68dc69494d50d586e8c0d841835e/packages/components/src/mobile/bottom-sheet/index.native.js#L307-L315) is called when opening the Inserter, which feels weird/wrong. That's not related to the embed block at all, but haven't tried this yet on `develop`. With the use of onDismiss in this PR, I worry their might be a bug hiding that could bite us. Can you have a look too? I will check develop in the meantime.

EDIT: ~confirmed that BlockSheet onDismiss() is called just by opening the Block Inserter on `develop` 🤔~. False alarm, what I was experiencing was the "add to start/add to end" sheet closing. It's too easy to trigger that sheet on the demo app due to the demo app being slow to process events.

fluiddot commented:
> I'm still in the process of reviewing the new changes @fluiddot , but wanted to surface something I noticed in the meantime: Looks like the BlockSheet [onDismiss()](https://github.com/WordPress/gutenberg/blob/6da0390f864c68dc69494d50d586e8c0d841835e/packages/components/src/mobile/bottom-sheet/index.native.js#L307-L315) is called when opening the Inserter, which feels weird/wrong. That's not related to the embed block at all, but haven't tried this yet on `develop`. With the use of onDismiss in this PR, I worry their might be a bug hiding that could bite us. Can you have a look too? I will check develop in the meantime.
> 
> EDIT: ~confirmed that BlockSheet onDismiss() is called just by opening the Block Inserter on `develop` 🤔~. False alarm, what I was experiencing was the "add to start/add to end" sheet closing. It's too easy to trigger that sheet on the demo app due to the demo app being slow to process events.

True, there's a bottom sheet ([code reference](https://github.com/WordPress/gutenberg/blob/014d837e85989208434b33918bfe0a200b5b64e7/packages/edit-post/src/components/layout/index.native.js#L172)) that is created along with other UI elements when the editor is mounted. I had the same issue time ago when I was debugging an issue related to the reusable blocks 😄 .

fluiddot committed:
Omit onDismiss prop from rest object

hypest commented:
✅ Add the Embed block
✅ Bottom sheet will open and link field will be focused
✅ If there is a link in the clipboard it should be auto-pasted
✅ Type a link if empty
✅ Can click on the “Learn more about embeds” link at the bottom to open the support article in the browser
✅ Close the bottom sheet
✅ If the link is not a valid URL, a warning message is shown

For the future additional work on the block, I think it would be nice if the "Invalid URL" is also displayed on the block itself somehow.

hypest commented:
> Missing features

We should probably add in the PR description that indeed we haven't implemented yet re-opening in the non-empty URL case, wdyt @fluiddot ?

fluiddot commented:
> > Missing features
> 
> We should probably add in the PR description that indeed we haven't implemented yet re-opening in the non-empty URL case, wdyt @fluiddot ?

Is that feature related to this item from the [referenced issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3274)?
> Can open by pressing the pen icon in the toolbar when an existing embed block is selected

If so, ~~I think the issue I should reference is [this one](https://github.com/wordpress-mobile/gutenberg-mobile/issues/3285)~~ looks like it's already referenced, or are you referring to a different case? 🤔

hypest commented:
Ah, sorry, I wasn't referring the the toolbar button which is indeed referenced in the Missing Features, but rather, to the fact that the current placeholder (when the URL _is_ set) is a text field that doesn't support tapping to edit/replace. I know that in the code and the PR comments that's clear, but I'm again imagining the future reader of the PR and we should message them that this PR doesn't support editing. Not a biggie though, we might as well just rephrase the reference to the pencil feature with something like "Editing is not supported yet and will be enabled when we'll implement the ...<pencil feature>". WDYT?

fluiddot commented:
> Ah, sorry, I wasn't referring the the toolbar button which is indeed referenced in the Missing Features, but rather, to the fact that the current placeholder (when the URL _is_ set) is a text field that doesn't support tapping to edit/replace. I know that in the code and the PR comments that's clear, but I'm again imagining the future reader of the PR and we should message them that this PR doesn't support editing. Not a biggie though, we might as well just rephrase the reference to the pencil feature with something like "Editing is not supported yet and will be enabled when we'll implement the ...". WDYT?

Oh ok, I get your point, it's true that once the URL is set it doesn't allow to replace it. I didn't thought about that because this feature is still under development but since we're going to merge it soon it's better to detail what's missing 👍 .

hypest reviewed (approved):
LGTM! Awesome work here @ceyhun and @fluiddot!

Thanks for picking this PR up and updating it @fluiddot while @ceyhun is on vacation. That way we can move the Embed port effort forward. Also, apologies for being a bit slow with the review from my side, or going down some darker corners a tad too much 🤗.

I think the implementation in this PR has turned out to be a bit complex, possibly because we are reusing most of `embed/edit.js` (a good approach anyway), and took me longer to wrap my head around it. I hope we left enough breadcrumbs around for our future selves to understand what's going on. We can always revisit and and more if needed 👍

Let's merge this!
